fastembed/common/model_management.py

```diff
@@ -107,36 +111,120 @@ def download_files_from_huggingface(
+            try:
+                for rel_path, meta in stored_metadata.items():
+                    file_path = model_dir / rel_path
+
+                    if not file_path.exists():
+                        return False
+
+                    if repo_files:  # online verification
+                        file_info = next((f for f in repo_files if f.path == file_path.name), None)
+                        if (
+                            not file_info
+                            or file_info.size != meta["size"]
+                            or file_info.blob_id != meta["blob_id"]
+                        ):
+                            return False
+
+                    else:  # offline verification
+                        if file_path.stat().st_size != meta["size"]:
+                            return False
+                return True
+            except (OSError, KeyError) as e:
+                logger.error(f"Error verifying files: {str(e)}")
+                return False
+
+        def _save_file_metadata(model_dir: Path, repo_files: list[RepoFile]) -> None:
+            try:
+                metadata = {}
+                file_info_map = {f.path: f for f in repo_files}
+
+                for file_path in model_dir.rglob("*"):
+                    if file_path.is_file() and file_path.name != cls.METADATA_FILE:
+                        repo_file = file_info_map.get(file_path.name)
+                        if repo_file:
+                            file_size = file_path.stat().st_size
+                            if file_size != repo_file.size:
+                                raise ValueError(f"File size mismatch for {file_path.name}")
+
+                            metadata[str(file_path.relative_to(model_dir))] = {
+                                "size": repo_file.size,
+                                "blob_id": repo_file.blob_id,
+                            }
+
+                if not metadata:  # happens when internet connection is down while downloading
+                    raise ValueError("No valid files found to save metadata")
+                else:
+                    (model_dir / cls.METADATA_FILE).write_text(json.dumps(metadata))
+            except (OSError, ValueError) as e:
+                logger.error(f"Error saving metadata: {str(e)}")
+                raise
+
         allow_patterns = [
             "config.json",
             "tokenizer.json",
             "tokenizer_config.json",
             "special_tokens_map.json",
             "preprocessor_config.json",
         ]
-        if extra_patterns is not None:
-            allow_patterns.extend(extra_patterns)
+
+        allow_patterns.extend(extra_patterns)
 
         snapshot_dir = Path(cache_dir) / f"models--{hf_source_repo.replace('/', '--')}"
-        is_cached = snapshot_dir.exists()
+        metadata_file = snapshot_dir / cls.METADATA_FILE
+
+        try:
+            # at least there's network first time (we should make sure to not fail if no network)
```

Comment:
not sure if I understand this comment, it looks like "at least there's network first time" is from the previous iteration